src/token/api/src/types/tx_record.rs

```diff
@@ -1,14 +1,23 @@
-use crate::types::{Operation, TransactionStatus, TxId};
 use candid::{CandidType, Deserialize, Principal};
 use ic_canister::ic_kit::ic;
 use ic_helpers::tokens::Tokens128;
 
+use crate::types::{Operation, TokenHolder, TokenReceiver, TransactionStatus, TxId};
+
+// `from/to` can be either a TokenHolder or a TokenReceiver or a Principal.
+#[derive(Deserialize, CandidType, Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub enum RecordOption {
+    TokenHolder(TokenHolder),
+    TokenReceiver(TokenReceiver),
+    Principal(Principal),
+}
+
 #[derive(Deserialize, CandidType, Debug, Clone)]
 pub struct TxRecord {
     pub caller: Option<Principal>,
     pub index: TxId,
-    pub from: Principal,
-    pub to: Principal,
+    pub from: RecordOption,
+    pub to: RecordOption,
```

Comment:
This seems strange to me. Can `to` field contain `RecordOption::TokenHolder` or `RecordOption::Principal`. I guess not. So the type for these fields is very misleading.

Reply:
Yes, We just made changes to the transfer methods only, for operations like `mint` and `burn` among others, these operation have unchanged `to` and `from` types (Principal) hence I made the enum for the transfer methods which takes `AccountIdentifier` and other methods which takes `Principal`.

Reply:
This looks like bad design to me. All the transactions now are always between account identifiers, not principals. So having the transactions history hold principals instead of account identifiers would be very confusing and difficult to work with. And also having `to` field containing `RecordOption::TokenHolder` instead of `RecordOption::TokenReceiver` whould be very bad.

Reply:
We don't have `subaccount` args for the mint/burn operation, as we're not taking them from the api, what I can do for now is create `Subaccount` as None and parse only the `Principal` operation as `Account Identifier` and save them as `Account Identifier` instead of `Principal`.
Also am going to use structs instead of enums, to enforce type checking for the `to` and `from` methods.